File: frontend/src/engine/capabilities/setRowColSize.ts
/**
 * setRowColSize – Set row height or column width manually.
 *
 * rowHeight is in points (default ~15).
 * columnWidth is in character widths (default ~8.43).
 */

import { CapabilityMeta, SetRowColSizeParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";

const meta: CapabilityMeta = {
  action: "setRowColSize",
  description: "Set row height or column width manually",
  mutates: false,
  affectsFormatting: true,
};

async function handler(
  context: Excel.RequestContext,
  params: SetRowColSizeParams,
  options: ExecutionOptions
): Promise<StepResult> {
  if (options.dryRun) {
    return { stepId: "", status: "success", message: `Would set ${params.dimension} to ${params.size} on ${params.range}` };
  }

  options.onProgress?.(`Setting ${params.dimension}...`);

  const sheet = context.workbook.worksheets.getActiveWorksheet();
  const range = sheet.getRange(params.range);

  if (params.dimension === "rowHeight") {
    range.format.rowHeight = params.size;
  } else {
    range.format.columnWidth = params.size;
  }

  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Set ${params.dimension} to ${params.size} on ${params.range}`,
    outputs: { range: params.range },
  };
}

registry.register(meta, handler as any);
export { meta };

File: backend/app/execution/capabilities/set_row_col_size.py
"""setRowColSize — set row height (points) or column width (character widths) on a range."""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    address = params.get("range")
    dimension = params.get("dimension")
    size = params.get("size")
    if not address or dimension not in ("rowHeight", "columnWidth") or size is None:
        return {
            "status": "error",
            "message": "setRowColSize requires 'range', 'dimension' (rowHeight|columnWidth), and 'size'.",
        }

    rng = resolve_range(ctx.workbook_handle, address)

    if ctx.dry_run:
        return {
            "status": "preview",
            "message": f"Would set {dimension} to {size} on {rng.address}.",
        }

    try:
        if dimension == "rowHeight":
            rng.row_height = float(size)
        else:
            rng.column_width = float(size)
    except Exception as exc:  # noqa: BLE001
        return {"status": "error", "message": f"Set size failed: {exc}", "error": str(exc)}

    return {
        "status": "success",
        "message": f"Set {dimension} to {size} on {rng.address}.",
        "outputs": {"range": rng.address, "dimension": dimension, "size": size},
    }


registry.register("setRowColSize", handler, mutates=False, affects_formatting=True)
